Adding arrays everywhere

katy-rex opened the pull request:
@tilikang This is a lot more than the two previous. I added arrays everywhere that I could.

The only place it didn't work like I thought it should is on the work.php page. The job descriptions would be long and I'm not sure what the best practices are to keep it readable.

Also, I'm imagining that each theatre will get their own public profile, so I created one of those and put it in a folder to keep stuff organized. But I think that was totally wrong since it absolutely doesn't work.

katy-rex committed:
Adding arrays everywhere

This is a lot more than the two previous. I added arrays everywhere that I could.

The only place it didn't work like I thought it should is on the work.php page. The job descriptions would be long and I'm not sure what the best practices are to keep it readable.

Also, I'm imagining that each theatre will get their own public profile, so I created one of those and put it in a folder to keep stuff organized. But I think that was totally wrong since it absolutely doesn't work.

katy-rex committed:
copy/paste error

tilikang commented on site/FAQ.php:
```diff
@@ -21,36 +21,18 @@
-				This is an answer for the question listed above.
-			</p>
+			<?php
+				$FAQ=array (
```
This looks like you've got the right idea in terms of how to make an array and loop through stuff, but why don't you take it to the next level. Eventually we're going to have a function called maybe `GetFaqQuestions()` which will pull from the database and return an array like this.

So how about you make a file called maybe faq.php which you include from init, and then make that a function which currently returns this hardcoded array? That way it'll be easy to replace the array with a database query later on, and this front-end code (the part that generates the html) won't need to be updated at all.

That's the beauty of functions. If you write them in a modular way, the code that uses them doesn't have to care what's happening behind the scenes. So switching from hardcoding arrays to using database queries is contained within a single function, and the rest of the code is blissfully unaware.

tilikang reviewed (commented):
@katy-rex - I left one comment here, but it applies to all these files.

Why don't you finish that and the header/footer changes from your other PR and then we'll merge everything into master. Once that's done, I feel like maybe you'll be ready to dive into mysql unless there's more you want to do in PHP first.

Schedule a 1:1 with me when you're ready, and if you're just killing time, I'd suggest you start reading some MySQL tutorials online.

katy-rex committed:
Merge branch 'header' into array

katy-rex committed:
Turning arrays into functions

katy-rex commented:
I put all the arrays into functions in separate files to include in init, and did some minor tweaks to quotations and spaces after "echo".

tilikang commented on site/FAQ.php:
```diff
@@ -1,41 +1,12 @@
-		</p>
+	<div class='Questions')>
+		<?php
+			GetFaqQuestions();
```
This is a step in the right direction, but even better would be to have this function return the array, but have the html printed in this page. The reason for this is that you might want to do different things with the FAQ questions array, and by printing html inside that function, you limit it to only helping for this one specific scenario.

Like in LACRM, think about how a `GetEvents` function might be used. It's used for the agenda on the Workspace, and agenda in the nav bar, the calendar view, and the agenda email that gets sent out each morning. All of those display the data in different ways, but they use the same underlying function for getting the events that need to be displayed.

So there's a single `GetEvents` function, and it returns raw data which is then echoed on this page.

Another angle to take when deciding what belongs in the function and what doesn't is to consider it like a front-end/back-end separation. The user actually loads the URL `FAQ.php` into their browser, and so this page is responsible for the front end (html, css, etc.). The files in the include folder are backend files which interact with the database, do complicated logic, etc. but they normally return data rather than echoing html directly.

Of course this somewhat contradicts what I already asked you to do with the header and footer function. Those are sort of exceptions because the exact same html will get printed from multiple different places. Using a function for that is ok, but you still want to separate the "front end" functions (e.g. `PrintFaqSection()`) from the "back end" functions (e.g. `GetFaqQuestions`). In that example, the front end function would call the back end function to get the raw data, but potentially other places might call that same front end function and use the data differently.

tilikang commented on site/config/init.php:
```diff
@@ -4,3 +4,8 @@
 include('include/db_query.php'); //this should happen right after config so other functions have access to the database
 include('include/header.php');
 include('include/footer.php');
+include('include/FAQFunction.php');
```
This is ok, but I want to make sure you're aware that it's normal to define multiple functions in one file. For example, you might just have a `header.php` file which could define the header and footer functions. These other files you've made are all different parts of the site so it's probably fine to have them as different files, but as you expand on this project, you'll want multiple functions in a single file. Like, `FAQFunction` might have a `CreateFaqItem` function, `GetFaqItem`, `GetAllFaqItems`, etc.

Side note: we didn't used to do it this way so you'll see examples in our main code base that don't follow this rule, but our current way of capitalizing things with acronyms like "FAW" is to only capitalize the first letter because otherwise it's hard to tell when the next word begins. So it'd be `FaqFunction` rather than `FAQFunction`.

katy-rex committed:
Updating arrays

Put the arrays in functions in separate files and kept the html on the front-end pages.

katy-rex commented:
That should all be updated now.
